### ISSUE TITLE
Clean up state listeners when a node is removed

### ISSUE THREAD

**Issue by KeatonTech** [emmanuelantony2000/valerie#7]
Currently state bindings to nodes are permanent, meaning that even if the node is removed from the DOM (user navigated to a different route, deleted the list item, closed the dialog, etc) it will still stay in memory and will still be updated whenever its state dependencies are updated. Naturally in some cases this could lead to excessive memory usage and slow page performance.

Here's my thinking on this issue, based on past experiences. Feel free to go a different way if you prefer.

Depending on the intended browser support, this might be a great use case for the new-ish [MutationObserver](https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver) API, which will execute a callback function whenever an element is removed from the DOM. I have heard of performance issues with this API in the past but they've hopefully gotten ironed out by now. If it's still too heavy, Valerie itself could track when State-bound elements are removed by the library itself, although this could lead to hard-to-find memory leaks when users circumvent Valerie and modify the DOM directly in their code.

The second issue is how to actually remove the state listeners. JS has no way of cancelling an `await` -- once a Promise is awaited, the calling `async` function will remain in memory (and take some CPU cycles) until that Promise gets resolved. So, this approach would work but have a potential memory and performance cost:

```rust
let mut old = StateId::new();
while let Some((new, value)) = stateRx.receive(old).await {
  if (element_was_removed(elem)) {
    break;
  }
  // ... perform element update ...
  old = new;
}
```

A safer solution would use a second channel that emits when the element is removed. Something like:

```rust
let mut old = StateId::new();
while true {
  let next_event = future::select(stateRx.receive(old), destroyedRx.receive(old)).await;
  match next_event {
      Either::Left((x, b)) =>{
        // ... perform element update ...
        old = new;
      },
      Either::Right((y, a)) => break,
  }
}
```

This will run the next loop iteration when either a) the state is updated or b) the element is removed. In practice I don't know if this actually ends up being more performant. It is safer in that elements that have been removed will no longer take up any resources, but heavier in that all state listeners must now listen to two different Futures.

Lastly there's the issue of where to store extra data about Nodes, in this case the destroyed state for each one (`destroyedRx` in the example above, or whatever boolean drives `element_was_removed` above that). As far as I can tell JsValues cannot be used as the key for a Set or HashMap, so I don't know how well a global datastore inside Valerie would work. That leaves storing the data in the DOM tree itself as the best option, either as HTML attributes or just as additional JS properties tacked on to the Node object. Both have pros and cons, and both specifically have the con that the stored values must all be JS-safe, eg. a Promise instead of a Future. There might be a hybrid approach that works, like storing an index somewhere on the JS object and then using that index as the key to a global datastore inside Rust. Honestly I have no idea which one would perform the best, I have a lot of JS development under my belt but I'm still fairly new to Rust (hopefully that isn't too obvious :) )

**Comment by emmanuelantony2000** [comment]
Firstly let me respond to the state listeners.

I don't think spawn in wasm-bindgen-futures interacts with JS, it spawns a future in Rust itself. As much as I have looked at the API, when the State variables go out of scope maybe removed from a list or any other case, even the tx goes out of scope and is destroyed. When the tx is destroyed, .await in rx will give a None value, and when it does the loop ends and the async function is over. This might be wrong so we can put a console.log like after the while loop and verify it.

If it is verified, we can have a remove listener function after the while loop and remove the listeners when the while loop ends.

Now the problem comes if we are passing the state variable inside the listener function, then it becomes some sort of a deadlock, so to prevent that maybe while passing it to an event listener we can pass a weak instead of an arc, so that it is guaranteed that the memory can be freed.

**Comment by KeatonTech** [comment]
Yeah I was trying to figure out how `spawn_local` actually works. My current theory is that it does use JS Promises since, as far as I can tell, Valerie doesn't pull in any Rust async runtime (like Tokio). I would actually hope it does use JS Promises, both because that's less code to import and also because in my experience JS async scheduling is more efficient than Rust scheduling (I think Promises don't use polling behind the scenes, and are instead purely function-based).

It's handy that Rx returns null when Tx gets destroyed, although that doesn't solve the whole problem. The state will often long outlive the components that bind to it. The fact that the components are owned by JS means that we can't implement `Drop` on them and so can't hook up a similar mechanism there.

**Comment by emmanuelantony2000** [comment]
Could we continue this conversation in discord if you don't mind. 😁

https://discord.gg/xx2sArF

**Comment by emmanuelantony2000** [comment]
So I did some tests with this piece of code.

```rust
use wasm_bindgen::prelude::*;
use wasm_timer::Delay;

#[wasm_bindgen]
extern {
    #[wasm_bindgen(js_namespace = console)]
    fn log(s: &str);
}

#[derive(Debug)]
struct Person(usize);

impl Drop for Person {
    fn drop(&mut self) {
        log(format!("{:?} is dropped.", self).as_str());
    }
}

#[wasm_bindgen(start)]
pub fn run() {
    let person1 = Person(1);
    log(format!("{:?} variable declared.", person1).as_str());

    let person2 = Person(2);
    log(format!("{:?} variable declared.", person2).as_str());

    log("First \'spawn_local\' going to be called.");
    wasm_bindgen_futures::spawn_local(async move {
        log(format!("{:?} is inside the async function now.", person1).as_str());
        log("The first async function is executing for 5 sec...");
        Delay::new(std::time::Duration::new(5, 0)).await.unwrap();
    });
    log("First \'spawn_local\' definition ended in the run() function.");

    log("Second \'spawn_local\' going to be called.");
    wasm_bindgen_futures::spawn_local(async move {
        log(format!("{:?} is inside the async function now.", person2).as_str());
        log("The second async function is executing for 5 sec...");
        Delay::new(std::time::Duration::new(5, 0)).await.unwrap();
    });
    log("Second \'spawn_local\' definition ended in the run() function.");

    log("End of run() function.");
}
```

And the console output obtained was:

```
11:37:54.426 Person(1) variable declared.
11:37:54.427 Person(2) variable declared.
11:37:54.427 First 'spawn_local' going to be called.
11:37:54.427 First 'spawn_local' definition ended in the run() function.
11:37:54.428 Second 'spawn_local' going to be called.
11:37:54.428 Second 'spawn_local' definition ended in the run() function.
11:37:54.428 End of run() function.
11:37:54.428 Person(1) is inside the async function now.
11:37:54.428 The first async function is executing for 5 sec...
11:37:54.428 Person(2) is inside the async function now.
11:37:54.429 The second async function is executing for 5 sec...
11:37:59.431 Person(1) is dropped.
11:37:59.433 Person(2) is dropped.
```

So indeed `Drop` works. And yes because you can't block wasm, so can't have an executor and for executing futures the approach taken is [here](https://github.com/rustwasm/wasm-bindgen/issues/1126).

I might be wrong here, but as much as I know, element doesn't get deleted from the DOM tree when it it removed, and to completely remove it along with event listeners, it has to be declared NULL. So if we use something like Rc and Refcell to enclose the FnMut before passing it, and then when the listener is dropped, Rc will automatically free the memory. I mean I have to yet try to implement it, but it may work out.

**Comment by emmanuelantony2000** [comment]
This issue was addressed and fixed in #8 so closing this.
